### PR TITLE
fix: Use non-slash separators when constructing relative PathTemplates [gax-php]

### PR DESCRIPTION
**Pull request by miraleung**

### CODE REVIEW

**Comment by miraleung** [comment]
@aohren FYI.

File: src/ResourceTemplate/Parser.php [hunk 1]
```diff
@@ -152,10 +152,11 @@ private static function parse($segmentString, $path, $index)
 
     /**
      * @param string $segmentStringWithoutBraces
+     * @param string separatorLiteral
```
Comment (dwsupplee) [inline src/ResourceTemplate/Parser.php:155]:
```suggestion
     * @param string $separatorLiteral
```

File: src/ResourceTemplate/RelativeResourceTemplate.php [hunk 2]
```diff
@@ -149,24 +155,73 @@ public function matches($path)
...
         // information about the variable identifier of any flattened segments. To do this, we
         // build a list of [string, Segment] tuples, where the string component is the appropriate
         // key.
         $keySegmentTuples = self::buildKeySegmentTuples($this->segments);
+
         $flattenedKeySegmentTuples = self::flattenKeySegmentTuples($keySegmentTuples);
         $flattenedKeySegmentTuplesCount = count($flattenedKeySegmentTuples);
         assert($flattenedKeySegmentTuplesCount > 0);
 
+
```
Comment (dwsupplee) [inline src/ResourceTemplate/RelativeResourceTemplate.php:175]: Would we please be able to drop this whitespace?

File: src/ResourceTemplate/RelativeResourceTemplate.php [hunk 3]
```diff
@@ -314,4 +369,22 @@ private static function countDoubleWildcards(array $segments)
         }
         return $doubleWildcardCount;
     }
+
+  /**
+   * Joins segments using their separators.
+   * @param array Segments.
```
Comment (dwsupplee) [inline src/ResourceTemplate/RelativeResourceTemplate.php:375]:
```suggestion
   * @param array $segmentsToRender
```

File: src/ResourceTemplate/RelativeResourceTemplate.php [hunk 4]
```diff
@@ -245,7 +289,7 @@ private function renderingException($bindings, $reason)
      * @param Segment[] $segments
      * @return array[] A list of [string, Segment] tuples
      */
-    private static function buildKeySegmentTuples(array $segments)
+    private static function buildKeySegmentTuples(array $segments, $separator = null)
```
Comment (dwsupplee) [inline src/ResourceTemplate/RelativeResourceTemplate.php:292]: Let's please add this to the docblock.

File: src/ResourceTemplate/Segment.php [hunk 5]
```diff
@@ -61,6 +61,10 @@ class Segment
     /** @var string */
     private $stringRepr;
 
+    /** @var string */
+    private $separator;
+
+
```
Comment (dwsupplee) [inline src/ResourceTemplate/Segment.php:67]: This extra whitespace can also be removed.

File: src/ResourceTemplate/Segment.php [hunk 6]
```diff
@@ -69,12 +73,18 @@ class Segment
...
      * @throws ValidationException
      */
-    public function __construct($segmentType, $value = null, $key = null, RelativeResourceTemplate $template = null)
-    {
+    public function __construct(
+        $segmentType,
+        $value = null,
+        $key = null,
+        RelativeResourceTemplate $template = null,
+        $separator = '/'
```
Comment (dwsupplee) [inline src/ResourceTemplate/Segment.php:81]: This should be added to the docblock.

File: src/ResourceTemplate/RelativeResourceTemplate.php [hunk 7]
```diff
@@ -277,7 +331,8 @@ private static function flattenKeySegmentTuples(array $keySegmentTuples)
                 case Segment::VARIABLE_SEGMENT:
                     // For segment variables, replace the segment with the segments of its children
                     $template = $segment->getTemplate();
-                    $nestedKeySegmentTuples = self::buildKeySegmentTuples($template->segments);
+                    $nestedKeySegmentTuples =
```
Comment (dwsupplee) [inline src/ResourceTemplate/RelativeResourceTemplate.php:334]: Just a minor style nit, but it if you are open to it, we generally prefer:

```php
$nestedKeySegmentTuples = self::buildKeySegmentTuples(
    $template->segments,
    $segment->getSeparator()
);
```

**Review by dwsupplee — CHANGES_REQUESTED** [review]
These comments are mostly style/doc nits, I will follow up and review the updated logic shortly. Thank you for this!

File: src/ResourceTemplate/Parser.php (hunk 1 above)
Comment (miraleung) [reply src/ResourceTemplate/Parser.php:155]: Done.

File: src/ResourceTemplate/RelativeResourceTemplate.php (hunk 2 above)
Comment (miraleung) [reply src/ResourceTemplate/RelativeResourceTemplate.php:175]: Done.

File: src/ResourceTemplate/RelativeResourceTemplate.php (hunk 3 above)
Comment (miraleung) [reply src/ResourceTemplate/RelativeResourceTemplate.php:375]: Done.

File: src/ResourceTemplate/RelativeResourceTemplate.php (hunk 4 above)
Comment (miraleung) [reply src/ResourceTemplate/RelativeResourceTemplate.php:292]: Done.

File: src/ResourceTemplate/Segment.php (hunk 5 above)
Comment (miraleung) [reply src/ResourceTemplate/Segment.php:67]: Done.

File: src/ResourceTemplate/Segment.php (hunk 6 above)
Comment (miraleung) [reply src/ResourceTemplate/Segment.php:81]: Done.

File: src/ResourceTemplate/RelativeResourceTemplate.php (hunk 7 above)
Comment (miraleung) [reply src/ResourceTemplate/RelativeResourceTemplate.php:334]: Done.

File: src/ResourceTemplate/RelativeResourceTemplate.php [hunk 8]
```diff
@@ -314,4 +371,22 @@ private static function countDoubleWildcards(array $segments)
         }
         return $doubleWildcardCount;
     }
+
+  /**
```
Comment (dwsupplee) [inline src/ResourceTemplate/RelativeResourceTemplate.php:375]: The indentation is off on this docblock

File: src/ResourceTemplate/RelativeResourceTemplate.php [hunk 9]
```diff
@@ -314,4 +371,22 @@ private static function countDoubleWildcards(array $segments)
...
+   * Joins segments using their separators.
+   * @param array $segmentsToRender.
```
Comment (dwsupplee) [inline src/ResourceTemplate/RelativeResourceTemplate.php:377]:
```suggestion
   * @param array $segmentsToRender
```

File: src/ResourceTemplate/RelativeResourceTemplate.php [hunk 10]
```diff
@@ -149,24 +155,72 @@ public function matches($path)
...
+                $rawPiece = substr($slashPathPieces[$pathPiecesIndex], $startIndex);
+                $pathPieceLength = strpos($rawPiece, $segment->getSeparator());
+                $pathPiece = substr($rawPiece, 0, $pathPieceLength);
+                $startIndex += $pathPieceLength + 1;
+            }
+            $pathPieces[] = $pathPiece;
+        }
+
+        if ($flattenedKeySegmentTuples[$i - 1][1]->getSegmentType() !== Segment::DOUBLE_WILDCARD_SEGMENT) {
+          // Process any remaining pieces. The binding logic will throw exceptions for any invalid paths.
```
Comment (dwsupplee) [inline src/ResourceTemplate/RelativeResourceTemplate.php:217]:
```suggestion
            // Process any remaining pieces. The binding logic will throw exceptions for any invalid paths.
```

**Review by dwsupplee — APPROVED** [review]
LGTM! Just a few more small tweaks then consider this ready to merge.

File: src/ResourceTemplate/RelativeResourceTemplate.php (hunk 10 above)
Comment (miraleung) [reply src/ResourceTemplate/RelativeResourceTemplate.php:217]: Done.

File: src/ResourceTemplate/RelativeResourceTemplate.php (hunk 9 above)
Comment (miraleung) [reply src/ResourceTemplate/RelativeResourceTemplate.php:377]: Done.

File: src/ResourceTemplate/RelativeResourceTemplate.php (hunk 8 above)
Comment (miraleung) [reply src/ResourceTemplate/RelativeResourceTemplate.php:375]: Done.

**Pull request merged**